udpa/config/v1alpha/routing.proto

```diff
@@ -0,0 +1,400 @@
+
+      oneof table_specifier {
+        // Inline match-action table.
+        InlineTable inline_table = 2;
+
+        // Support VHDS-like behavior, on-demand fetching of header routing
+        // tables. The resource type is Action, the resource name is the string
+        // value match to be used. E.g. an Action called "*-foo.com" may be
+        // delivered dynamically. This can't be used with regex match
+        // descriptors.
+        ConfigProvider config_provider = 3;
+      }
+    }
+
+    // Match-action table for integer ranges.
+    message IntegerRangeMatcher {
+      message IntegerRangeMatchAction {
+        Int64Range range = 1;
+        Action action = 2;
+      }
+      // Ranges must not overlap.
+      repeated IntegerRangeMatchAction match_actions = 1;
+    }
+
+    oneof header_match_specifier {
+      // String value match-action.
+      StringMatcher string_matcher = 3;
+
+      // Integer value match-action.
+      IntegerRangeMatcher integer_range_matcher = 4;
+    }
+  }
+
+  // Currently only header matchers are supported.
+  oneof matcher_specifier {
+    option (validate.required) = true;
+    HeaderMatcher header_matcher = 1;
+  }
+
+  // Fall-thru action if no other match occurs.
+  Action default_action = 2;
+}
+
+// A linear matcher provides the ability to iterate over a list of (match
+// criteria, action) pairs, where the match criteria are a set of header
+// matchers.
+message LinearMatcher {
+  message MatchAction {
+    message HeaderMatcher {
+      // Specifies the name of the header in the request.
+      string name = 1 [(validate.rules).string.min_bytes = 1];
+
+      // If specified, the match result will be inverted before checking.
+      // Defaults to false.
+      //
+      // Examples:
+      //
+      // * The regex *\d{3}* does not match the value *1234*, so it will match
+      //   when inverted.
+      // * The range [-10,0) will match the value -1, so it will not match when
+      //   inverted.
+      bool invert_match = 2;
+
+      // How the header value is extracted for match.
+      HeaderValueExtractor value_extractor = 3;
+
+      // Matcher for string values.
+      message StringMatcher {
+        StringMatchDescriptor match_descriptor = 1;
+        string value = 2;
+      }
+
+      // Matcher for integer ranges.
+      message IntegerRangeMatcher {
+        Int64Range range = 1;
+      }
+
+      oneof header_matcher_specifier {
+        option (validate.required) = true;
+
+        StringMatcher string_matcher = 4;
+
+        IntegerRangeMatcher integer_range_matcher = 5;
+      }
+    }
+
+    // List of header matchers providing the match criteria. All HeaderMatchers
+    // must match on the request; a logical AND of the HeaderMatcher results is
+    // performed.
+    repeated HeaderMatcher header_matchers = 1;
+
+    // Action to take on match of all header_matchers.
+    Action action = 2;
+  }
+  repeated MatchAction match_actions = 1;
+}
+
+message Matcher {
```

Comment:
It's worth noting that the current xDS data model actually uses these matchers for two things:

1. Determining configuration parameters for things like timeouts and retries.
2. Determining how to route requests (e.g., which cluster to send the request to).

Note that routing could actually be done by the LB policy instead of being done here.  If we wind up going with the extensible hierarchical LB config approach that @snowp and I are discussing in envoyproxy/envoy#5598, then this would even provide a clean and extensible way to choose between linear and hierarchical matching -- we could just implement each level of the hierarchical match in its own LB policy, so that we would not even need a separate configuration structure for linear vs. hierarchical matching.

I think that we would still need *some* matching criteria here for determining configuration parameters for things like timeouts and retries.  But if we push most of this complexity into the LB policy, can we make what remains here much simpler?  I am wondering if we can live with just some very basic criteria here, such as prefix path matching only, to select the configuration.  This would allow a bit less flexibility in selecting the configuration parameters, but it would make the routing much more extensible, and it would address a fairly signficant mismatch between the xDS data model and the gRPC architecture that is causing some friction.

We ran into this just recently as we started looking into supporting some of this matching in gRPC, and we quickly realized that we have a bit of a chicken-and-egg problem: we handle config selection via a resolver plugin, which does not have access to per-request data, and we handle request routing via an LB policy plugin, which runs after config selection has already been performed.  For now, we're going to handle this by supporting only very simple path-prefix matching, which is essentially the mechanism we currently use to select configs on a [per-method basis](https://github.com/grpc/grpc-proto/blob/38fb94f83af4dcb5229e171f5a0342846d4a5f2f/grpc/service_config/service_config.proto#L39).  But because the current xDS data model ties the config selection and routing matching together, this means that we can't support things like regexp matching or matching on headers, which we could otherwise support for request routing.  (There are a number of reasons why we don't want to support more complex matching for config selection, mainly related to performance and complexity.)  If we could decouple these two things, then we could support the more flexible matching for request routing.

So I guess the question here is, could we make the matching criteria here simpler if they were going to be used only for selecting configuration parameters like timeouts and retries?

Reply:
I can see the appeal behind such an approach, but I worry that it's too big of a change given how Envoy operates today: for HTTP proxies today each filter has access to the routing decision (i.e. which routing action to  make,  which cluster to target, etc.) and is able to act on that. 

The load balancing choice today happens at the terminal filter, and the interface essentially just returns a handle to the upstream host to connect to. By pushing more of what is currently expressed in the  route table into the load balancer none of that information would be made available to the various filters. 

Now, it might be feasible to split routing and config selection apart without pushing it into the load balancer. I imagine we could have two separate resolution steps,  one for determining  the config and one for routing. That way  on the Envoy side both results could be cached as the route selection is today, and gRPC would able to implement each resolver where it makes the most sense.